Number utils

blanchco opened the pull request:
# Description

1. Created some functions to do some conversions for the new input field:
- number -> exponential
- exponential -> number
- number -> NIST
- NIST -> number
- displaying a number (this depends on the length)

2. I have made the input for all these functions a string, since we can deal with very small or very large numbers it can lead to precision issues if we were to use the Number library to manipulate these inputs

3. UNIT TESTS!!

mwdchang commented on packages/client/hmi-client/src/utils/number.ts:
```diff
@@ -0,0 +1,159 @@
+ * @param {string} num - The number string to clean.
+ * @returns {string} The cleaned number string.
+ */
+export function cleanNumberString(num: string): string {
```
Why not just use the native `parseFloat`, since what you are doing is pretty much, well, parsing a float. 

The native implementation has additional features to deal with scientific notations and other things, which yours does not.

blanchco replied:
got rid of this function

mwdchang commented on packages/client/hmi-client/src/utils/number.ts:
```diff
@@ -0,0 +1,159 @@
+	}
+	// Remove spaces from the integer part
+	return numStr.replace(/ /g, '');
+}
```
Another case of `parseFloat`?

```
parseFloat(`num.replace(/ /g, ''))
```

blanchco replied:
this now uses `parseFloat`. I still need to scrub all the whitespace as `parseFloat` stops parsing at the first instance of a whitespace, while a NIST number can have many whitespaces inside it

mwdchang commented:
With regards to **point-2,** I am a bit confused if you are referring to floating-point precision, or if you are talking about allowing arbitrary precision.

- If the former, string vs number will not solve anything, the problem is inherently baked into binary representation. Short of using a specialized library we just need to be careful and deal with it. e.g. `0.2 * 0.2 != 0.04`
- if the latter, I don't think we need to worry too much about it. Our problem domains should not have extremely large numbers or exceedingly small numbers. 

So overall it would be good if you elaborate a bit more on the design goals.

blanchco commented:
@mwdchang I've address these comments and am focusing on formatting rather than parsing.

Change made:
removed `cleanNumberString`
removed `numberToExponential`, we can just use `toExponential`
`exponentialToNumber` now uses toLocaleString to properly display the full number rather than the custom parsing
`numberToNist` is relatively unchanged aside from using parseFloat to clean the number input
`nistToNumber` cleans _all_ whitespaces and then parseFloat rather than the custom parsing

blanchco commented on packages/client/hmi-client/src/utils/number.ts:
```diff
@@ -0,0 +1,159 @@
+	const zeros = '0'.repeat(mag);
+	console.log(sign + str + zeros);
+	return sign + str + zeros;
+}
```
rewrote this function to use parseFloat

liunelson commented on packages/client/hmi-client/tests/unit/utils/number.spec.ts:
```diff
@@ -0,0 +1,126 @@
+		});
+	});
+
+	describe('numberToNist', () => {
```
Need example here and `nistToNumber` and `displayNumber` for this rule
![image](https://github.com/DARPA-ASKEM/terarium/assets/1030720/f123350a-ea6a-48c1-9215-d2eba0a76cb5)

>However, when there are only four digits before or after the decimal marker, we recommend that no space is required and no space should be used

For example:
`1234.123` -> `1234.123`
`1234.1234` -> `1234.1234`
`11234.123` -> `11 234.123`
`1234.12344` -> `1 234.123 44`
`11234.123` -> `11 234.123`
`11234.1234` -> `11 234.123 4`

liunelson commented on packages/client/hmi-client/tests/unit/utils/number.spec.ts:
```diff
@@ -0,0 +1,126 @@
+import { exponentialToNumber, numberToNist, nistToNumber, displayNumber } from '@/utils/number';
+
+describe('number util tests', () => {
+	describe('exponentialToNumber', () => {
```
Probably should handle this number with thin spaces and exponentials
e.g. `-1 234.567 89e+1` -> `-1.234 567 89e+4` -> `-12345.6789`
